File: lib/src/main/kotlin/org/wireforce/virbet/retrofit/RetrofitInterfaceMain.java
package org.wireforce.virbet.retrofit;

import kotlin.Suppress;
import org.wireforce.virbet.dto.*;
import retrofit2.Call;
import retrofit2.http.*;

/**
 * Интерфейс `RetrofitInterfaceMain` содержит объявления методов для взаимодействия с удаленным сервером
 * посредством библиотеки Retrofit в контексте основных операций приложения.
 *
 * <p><b>Пример использования:</b></p>
 * <pre>{@code
 * RetrofitInterfaceMain retrofitInterface = Retrofit.create(RetrofitInterfaceMain.class);
 * Call<KtorResponse<KtorIdKeyRelease>> call = retrofitInterface.generateAnonymousIDKey("AnyString");
 * KtorResponse<KtorIdKeyRelease> response = call.execute().body();
 * }</pre>
 */
interface RetrofitInterfaceMain {
    /**
     * Выполняет GET-запрос для получения ID ключа.
     * ID ключ предназначен для представления пользователя без логина и пароля,
     * обеспечивая анонимный вход в систему.
     *
     * @param requestId Строка, представляющая идентификатор запроса (по умолчанию генерируется случайным образом).
     *                  Служит для предотвращения безконтрольного выпуска ключей.
     *                  По умолчанию пользователь может выпустить 3 ID ключа на один IP-адрес в целях безопасности,
     *                  но существует вероятность холостого выпуска ключей. Чтобы предотвратить такой выпуск, используйте requestId.
     *                  ВНИМАНИЕ! requestId кешируется на 24 секунды с момента первого запроса.
     *
     *
     *                  <p><b>Пример использования:</b></p>
     *
     *                  <pre>{@code
     *                  Call<KtorResponse<KtorIdKeyRelease>> call = retrofitInterface.generateAnonymousIDKey("AnyString");
     *                  KtorResponse<KtorIdKeyRelease> response = call.execute().body();
     *                  }</pre>
     * @return Объект типа {@code Call<KtorResponse<KtorIdKeyRelease>>}, представляющий асинхронный результат запроса.
     * * @see #login(KtorAuthThoughIdKey)
     */
    @GET("/auth/requestIdKey")
    @Suppress(names = "Unused")
    Call<KtorResponse<KtorIdKeyRelease>> generateAnonymousIDKey(
            @Query("requestId") String requestId
    );

    /**
     * Метод {@code login} выполняет POST-запрос для входа в систему через ключ ID.
     *
     * @param idKey Строка, представляющая тело запроса.
     *              <p>
     *              Пример использования:
     *              <pre>{@code
     *                           KtorAuthThoughIdKey idKey = new KtorAuthThoughIdKey("your_id_key_here");
     *                           Call<KtorResponse<KtorLoginInUser>> call = retrofitInterfaceMain.login(idKey);
     *                           call.enqueue(callback);
     *                           }</pre>
     * @return Объект типа {@code Call<KtorResponse<KtorLoginInUser>>}, представляющий асинхронный результат запроса.
     * @see #generateAnonymousIDKey(String)
     * @since 1
     */
    @POST("/auth/internal/loginThoughIDKey")
    @Suppress(names = "Unused")
    Call<KtorResponse<KtorIdKeyRelease>> login(@Body KtorAuthThoughIdKey idKey);

    /**
     * Метод {@code login} выполняет POST-запрос для входа пользователя.
     *
     * @param body Объект типа {@code KtorLoginUser}, содержащий данные для входа.
     *             <p>
     *             Пример использования:
     *             <pre>{@code
     *                         KtorLoginUser userCredentials = new KtorLoginUser("your_username", "your_password");
     *                         Call<KtorResponse<KtorLoginInUser>> call = retrofitInterfaceMain.login(userCredentials);
     *                         call.enqueue(callback);
     *                         }</pre>
     */
    @POST("/auth/internal/login")
    @Suppress(names = "Unused")
    Call<KtorResponse<KtorLoginInUser>> login(@Body KtorLoginUser body);


    /**
     * Метод {@code signup} выполняет POST-запрос для регистрации нового пользователя.
     *
     * @param body Объект типа {@code KtorCreateUser}, содержащий данные для регистрации пользователя.
     *
     *             <p>
     *             Пример использования:
     *             <pre>{@code
     *                         KtorCreateUser newUser = new KtorCreateUser("new_username", "new_password");
     *                         Call<KtorResponse<KtorCreatedUser>> call = retrofitInterfaceMain.signup(newUser);
     *                         call.enqueue(callback);
     *                         }</pre>
     * @return Объект типа {@code Call<KtorResponse<KtorCreatedUser>>}, представляющий асинхронный результат запроса.
     */
    @POST("/auth/internal/signup")
    @Suppress(names = "Unused")
    Call<KtorResponse<KtorCreatedUser>> signup(@Body KtorCreateUser body);

    /**
     * Метод {@code getMe} выполняет GET-запрос для получения информации о текущем пользователе.
     *
     * @param jwtBearerToken Строка, содержащая JWT-токен, используемый для аутентификации.
     *
     *                       <p>
     *                       Пример использования:
     *                       <pre>{@code
     *                                             String jwtToken = "your_jwt_token_here";
     *                                             Call<KtorResponse<KtorObtainMe>> call = retrofitInterfaceMain.getMe(jwtToken);
     *                                             call.enqueue(callback);
     *                                             }</pre>
     * @return Объект типа {@code Call<KtorResponse<KtorObtainMe>>}, представляющий асинхронный результат запроса.
     */
    @GET("/account/me")
    @Headers("Content-Type: application/json;charset=UTF-8")
    @Suppress(names = "Unused")
    Call<KtorResponse<KtorObtainMe>> getMe(@Header("Authorization") String jwtBearerToken);

    /**
     * Метод {@code getBalance} выполняет GET-запрос для получения баланса текущего пользователя.
     *
     * @param jwtBearerToken Строка, содержащая JWT-токен, используемый для аутентификации.
     *
     *                       <p>
     *                       Пример использования:
     *                       <pre>{@code
     *                                             String jwtToken = "your_jwt_token_here";
     *                                             Call<KtorResponse<KtorObtainBalance>> call = retrofitInterfaceMain.getBalance(jwtToken);
     *                                             call.enqueue(callback);
     *                                             }</pre>
     * @return Объект типа {@code Call<KtorResponse<KtorObtainBalance>>}, представляющий асинхронный результат запроса.
     */
    @GET("/wallet/balance")
    @Suppress(names = "Unused")
    @Headers("Content-Type: application/json;charset=UTF-8")
    @Deprecated()
    Call<KtorResponse<KtorObtainBalance>> getBalance(@Header("Authorization") String jwtBearerToken);

    /**
     * Метод {@code createBalanceIfNotExists} выполняет POST-запрос для создания баланса текущего пользователя,
     * если его еще не существует.
     *
     * @param jwtBearerToken Строка, содержащая JWT-токен, используемый для аутентификации.
     *
     *                       <p>
     *                       Пример использования:
     *                       <pre>{@code
     *                                             String jwtToken = "your_jwt_token_here";
     *                                             Call<KtorResponse<KtorCreatedBalance>> call = retrofitInterfaceMain.createBalanceIfNotExists(jwtToken);
     *                                             call.enqueue(callback);
     *                                             }</pre>
     * @return Объект типа {@code Call<KtorResponse<KtorCreatedBalance>>}, представляющий асинхронный результат запроса.
     */
    @POST("/wallet/createBalance")
    @Suppress(names = "Unused")
    @Headers("Content-Type: application/json;charset=UTF-8")
    @Deprecated()
    Call<KtorResponse<KtorCreatedBalance>> createBalanceIfNotExists(@Header("Authorization") String jwtBearerToken);

    /**
     * Метод {@code getBalanceOrCreate} выполняет GET-запрос для получения баланса текущего пользователя
     * или создания его, если его еще не существует.
     *
     * @param jwtBearerToken Строка, содержащая JWT-токен, используемый для аутентификации.
     *
     *                       <p>
     *                       Пример использования:
     *                       <pre>{@code
     *                                             String jwtToken = "your_jwt_token_here";
     *                                             Call<KtorResponse<KtorCreatedBalance>> call = retrofitInterfaceMain.getBalanceOrCreate(jwtToken);
     *                                             call.enqueue(callback);
     *                                             }</pre>
     * @return Объект типа {@code Call<KtorResponse<KtorCreatedBalance>>}, представляющий асинхронный результат запроса.
     */
    @GET("/wallet/balanceGetOrCreate")
    @Suppress(names = "Unused")
    @Headers("Content-Type: application/json;charset=UTF-8")
    Call<KtorResponse<KtorCreatedBalance>> getBalanceOrCreate(@Header("Authorization") String jwtBearerToken);
}

File: lib/src/main/kotlin/org/wireforce/virbet/retrofit/RetrofitInterfaceMain.kt
package org.wireforce.virbet.retrofit

import com.aventrix.jnanoid.jnanoid.NanoIdUtils
import org.jetbrains.annotations.NotNull
import org.wireforce.virbet.classes.Constants
import org.wireforce.virbet.classes.ContentLanguage
import org.wireforce.virbet.classes.Sport
import org.wireforce.virbet.dto.*
import retrofit2.Call
import retrofit2.http.*


/**
 * Интерфейс `RetrofitInterfaceMain` содержит объявления методов для взаимодействия с удаленным сервером
 * посредством библиотеки Retrofit в контексте основных операций приложения.
 *
 */
interface RetrofitInterfaceMain {
	/**
	 * Метод `generateAnonymousIDKey` выполняет GET-запрос для получения ID ключа.
	 * Данный ID ключ служит для представления пользователя без логина и пароля,
	 * обеспечивая анонимный вход в систему.
	 *
	 * @param requestId Строка, представляющая идентификатор запроса (по умолчанию генерируется случайным образом).
	 *                  Служит для предотвращения безконтрольного выпуска ключей.
	 *                  По умолчанию пользователь может выпустить 3 ID ключа на один IP-адрес в целях безопасности,
	 *                  но существует вероятность холостого выпуска ключей. Чтобы предотвратить такой выпуск, используйте requestId.
	 *                  ВНИМАНИЕ! requestId кешируется на 24 секунды с момента первого запроса
	 *
	 * @return Объект типа `Call<KtorResponse<KtorIdKeyRelease>>`, представляющий асинхронный результат запроса.
	 * @since 1
	 * @see login
	 *
	 * Примеры использования:
	 * - requestIdKey() // Возвращает ID ключа, например, "JHGKHJDGKJHG"
	 * - requestIdKey() // Возвращает новый ID ключа, например, "VLGKAJHGKJDD"
	 * - requestIdKey("AnyString") // Возвращает ID ключа, например, "JHGKHJDGKJHG"
	 * - requestIdKey("AnyString") // Возвращает тот же ID ключа, что и предыдущий вызов, так как использован requestId
	 */
	@GET("/auth/requestIdKey")
	@Suppress("Unused")
	fun generateAnonymousIDKey(
		@Query("requestId") requestId: String = NanoIdUtils.randomNanoId()
	): Call<KtorResponse<KtorIdKeyRelease>>

	/**
	 * Метод `login` выполняет POST-запрос для входа в систему через ключ ID.
	 *
	 * @since 1
	 * @see generateAnonymousIDKey
	 * @param idKey Строка, представляющая тело запроса.
	 * @return Объект типа `Call<KtorResponse<KtorLoginInUser>>`, представляющий асинхронный результат запроса.
	 */
	@POST("/auth/internal/loginThoughIDKey")
	@Suppress("Unused")
	fun login(@Body idKey: KtorAuthThoughIdKey): Call<KtorResponse<KtorLoginInUser>>

	/**
	 * Метод `login` выполняет POST-запрос для входа пользователя.
	 *
	 * @param body Объект типа `KtorLoginUser`, содержащий данные для входа.
	 */
	@POST("/auth/internal/login")
	@Suppress("Unused")
	fun login(@Body body: KtorLoginUser): Call<KtorResponse<KtorLoginInUser>>

	/**
	 * Метод `signup` выполняет POST-запрос для регистрации нового пользователя.
	 *
	 * @param body Объект типа `KtorCreateUser`, содержащий данные для регистрации пользователя.
	 * @return Объект типа `Call<KtorResponse<KtorCreatedUser>>`, представляющий асинхронный результат запроса.
	 */
	@POST("/auth/internal/signup")
	@Suppress("Unused")
	fun signup(@Body body: KtorCreateUser): Call<KtorResponse<KtorCreatedUser>>

	/**
	 * Метод `getMe` выполняет GET-запрос для получения информации о текущем пользователе.
	 *
	 * @param jwtBearerToken Строка, содержащая JWT-токен, используемый для аутентификации.
	 * @return Объект типа `Call<KtorResponse<KtorObtainMe>>`, представляющий асинхронный результат запроса.
	 */
	@GET("/account/me")
	@Suppress("Unused")
	@Headers("Content-Type: application/json;charset=UTF-8")
	fun getMe(@Header("Authorization") jwtBearerToken: String): Call<KtorResponse<KtorObtainMe>>

	@GET("/wallet/balance")
	@Suppress("Unused")
	@Headers("Content-Type: application/json;charset=UTF-8")
	fun getBalance(@Header("Authorization") jwtBearerToken: String): Call<KtorResponse<KtorObtainBalance>>

	@POST("/wallet/createBalance")
	@Suppress("Unused")
	@Headers("Content-Type: application/json;charset=UTF-8")
	fun createBalanceIfNotExists(@Header("Authorization") jwtBearerToken: String): Call<KtorResponse<KtorCreatedBalance>>

	@GET("/wallet/balanceGetOrCreate")
	@Suppress("Unused")
	@Headers("Content-Type: application/json;charset=UTF-8")
	fun getBalanceOrCreate(@Header("Authorization") jwtBearerToken: String): Call<KtorResponse<KtorCreatedBalance>>

	/**
	 * Метод `getRelevantEvents` выполняет GET-запрос для получения актуальных событий.
	 *
	 * @param jwtBearerToken Строка, содержащая JWT-токен, используемый для аутентификации.
	 * @param requiredContentLanguage Язык контента (на каком языке должен быть контент с событиями) типа `ContentLanguage` (по умолчанию `ContentLanguage.EN`).
	 *
	 * @param sportId Идентификатор вида спорта типа `Sport`. Аннотация `@NotNull` указывает, что параметр не может быть `null`.
	 * @param limit Максимальное количество событий (по умолчанию значение из Constants.RETROFIT_DEFAULT_ZIPPED_LIMIT).
	 * @param offset Смещение (по умолчанию значение из Constants.RETROFIT_DEFAULT_ZIPPED_OFFSET).
	 * @return Объект типа `Call<KtorResponse<KtorEventsRequest<KtorParentEventPreview>>>`, представляющий асинхронный результат запроса.
	 */
	@GET("/events/relevant")
	@Suppress("Unused")
	@Headers("Content-Type: application/json;charset=UTF-8")
	fun getRelevantEvents(
		@Header("Authorization") jwtBearerToken: String,
		@Header("Content-Language") requiredContentLanguage: ContentLanguage = ContentLanguage.EN,
		@Query("sportId") @NotNull sportId: Sport,
		@Query("limit") limit: Int = Constants.RETROFIT_DEFAULT_ZIPPED_LIMIT,
		@Query("offset") offset: Int = Constants.RETROFIT_DEFAULT_ZIPPED_OFFSET,
	): Call<KtorResponse<KtorEventsRequest<KtorEventSlim>>>
}